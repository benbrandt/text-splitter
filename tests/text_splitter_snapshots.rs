use std::fs;

use more_asserts::assert_le;
use once_cell::sync::Lazy;
use text_splitter::{Characters, ChunkSizer, TextSplitter};
use tiktoken_rs::{cl100k_base, CoreBPE};
use tokenizers::Tokenizer;

#[test]
fn characters_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default();
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(Characters.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn characters_range() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        let splitter = TextSplitter::default();
        let chunks = splitter.chunks(&text, 500..=2000).collect::<Vec<_>>();

        assert_eq!(chunks.join(""), text);
        for chunk in chunks.iter() {
            assert_le!(Characters.chunk_size(chunk), 2000);
        }
        insta::assert_yaml_snapshot!(chunks);
    });
}

#[test]
fn characters_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default().with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static HUGGINGFACE_TOKENIZER: Lazy<Tokenizer> =
    Lazy::new(|| Tokenizer::from_pretrained("bert-base-cased", None).unwrap());

#[test]
fn huggingface_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(HUGGINGFACE_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn huggingface_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static TIKTOKEN_TOKENIZER: Lazy<CoreBPE> = Lazy::new(|| cl100k_base().unwrap());

#[test]
fn tiktoken_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(TIKTOKEN_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn tiktoken_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}
