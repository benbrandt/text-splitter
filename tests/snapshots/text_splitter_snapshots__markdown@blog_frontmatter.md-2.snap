---
source: tests/text_splitter_snapshots.rs
expression: chunks
input_file: tests/inputs/markdown/blog_frontmatter.md
---
- "---\nlayout: page\ntitle: Philosophy\npermalink: /philosophy/\n---"
- "\n\n"
- "Jekyll offers a unique philosophy when approaching the problem of static\nsite generation. "
- "This core philosophy drives development and product\n"
- "decisions. When a contributor, maintainer, or user asks herself what Jekyll\n"
- "is about, the following principles should come to mind:\n\n"
- "### 1. No Magic\n\n"
- "Jekyll is not magic. A user should be able to understand the underlying\n"
- "processes that make up the Jekyll build without much reading. It should\n"
- "do only what you ask it to and nothing more. When a user takes a certain\n"
- "action, the outcome should be easily understandable and focused.\n\n"
- "### 2. It \"Just Works\"\n\n"
- "The out-of-the-box experience should be that it \"just works.\" Run\n"
- "`gem install jekyll`"
- " and it should build any Jekyll site that it's given.\n"
- "Features like auto-regeneration and settings like the markdown renderer\n"
- "should represent sane defaults that work perfectly for the vast majority of\n"
- "cases. The burden of initial configuration should not be placed on the user.\n\n"
- "### 3. Content is King\n\n"
- "Why is Jekyll so loved by content creators? It focuses on content first and\n"
- "foremost, making the process of publishing content on the Web easy. Users\n"
- "should find the management of their content enjoyable and simple.\n\n"
- "### 4. Stability\n\n"
- "If a user's site builds today, it should build tomorrow.\n"
- "Backwards-compatibility should be strongly preferred over breaking changes.\n"
- "Breaking changes should be made to support a strong practical goal, and\n"
- "breaking changes should never be made to drive forward \"purity\" of the\n"
- "codebase, or other changes purely to make the maintainers' lives easier.\n"
- "Breaking changes provide a significant amount of friction between upgrades\n"
- "and reduce the confidence of users in this software, and should thus be\n"
- "avoided unless absolutely necessary.\n"
- "Upon breaking changes, provide a clear path for users to upgrade.\n\n"
- "### 5. Small & Extensible\n\n"
- "The core of Jekyll should be simple and small, and extensibility should be\n"
- "a first-class feature to provide added functionality from community\n"
- "contributors. The core should be kept to features used by at least 90% of\n"
- "usersâ€“everything else should be provided as a plugin. New features should\n"
- "be shipped as plugins and focus should be put on creating extensible core\nAPI'"
- "s to support rich plugins.\n"
